### PR TITLE
RFC FS-1006 Struct Tuples and Interop with C# 7.0 Tuples

### PR DESCRIPTION
**Pull request by dsyme**
This is the WIP PR for [F# RFC FS-1006  Struct Tuples and Interop with C# 7.0 Tuples](https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1006-struct-tuples.md)

Please discuss the implementation on this thread, and the RFC and design on the RFC discussion thread.

### CODE REVIEW

**Comment by dsyme** [comment]
Current failures are all to do with the added FSHarp.Core surface area.  This will disappear once we can use a System.ValueTuples.dll from .NET

[00:21:11] 1) Failed : FSharp.Core.Unittests.SurfaceArea.SurfaceAreaTest.VerifyArea
[00:21:11] 2) Failed : FSharp.Core.Unittests.Portable.SurfaceArea.SurfaceAreaTest.VerifyArea
[00:21:11] 3) Failed : FSharp.Core.Unittests.Portable.SurfaceArea.SurfaceAreaTest.VerifyArea
[00:21:11] 4) Failed : FSharp.Core.Unittests.Portable.SurfaceArea.SurfaceAreaTest.VerifyArea
[00:21:11] 5) Failed : FSharp.Core.Unittests.Portable.SurfaceArea.SurfaceAreaTest.VerifyArea

File: src/fsharp/TastOps.fs [hunk 1]
```diff
@@ -202,6 +225,21 @@ and remapMeasureAux tyenv unt =
           else unt
        | Some (TType_measure unt) -> remapMeasureAux tyenv unt
        | Some ty -> failwithf "incorrect kinds: %A" ty
+
+and remapTupInfoAux _tyenv unt =
+    match unt with
+    | TupInfo.Const _ -> unt
+    //| TupInfo.Var tp as unt ->
```
Comment (TIHan) [inline src/fsharp/TastOps.fs]: Why is all this commented out?

File: src/fsharp/TastOps.fs [hunk 2]
```diff
@@ -635,6 +674,20 @@ let rec stripTyEqnsA g canShortcut ty =
 
 let stripTyEqns g ty = stripTyEqnsA g false ty
 
+let rec stripTupInfoEqns aexpr = 
+    match aexpr with 
+    //| TupInfo.Var v ->
```
Comment (TIHan) [inline src/fsharp/TastOps.fs]: Similar to previous comment elsewhere. Just wondering why some things are commented out?

File: src/fsharp/TastOps.fs [hunk 3]
```diff
@@ -5089,8 +5169,13 @@ let rec tyOfExpr g e =
         | TOp.ExnConstr _ -> g.exn_ty
         | TOp.Bytes _ -> mkByteArrayTy g
         | TOp.UInt16s _ -> mkArrayType g g.uint16_ty
+<<<<<<< HEAD
```
Comment (TIHan) [inline src/fsharp/TastOps.fs]: I think a merge got screwed up here.

File: src/fsharp/tast.fs [hunk 4]
```diff
@@ -3277,39 +3283,57 @@ and
...
 
-and MeasureExpr = 
+and [<RequireQualifiedAccess>] TupInfo = 
+    // /// A variable tupInfo
+    // | Var of Typar
+
+    /// Some constant, e.g. true or false for tupInfo
+    | Const of bool
+
+and [<RequireQualifiedAccess>] Measure =
```
Comment (TIHan) [inline src/fsharp/tast.fs]: This is much nicer than before! :)

File: src/utils/sformat.fs [hunk 5]
```diff
@@ -296,45 +296,6 @@ namespace Microsoft.FSharp.Text.StructuredFormat
             (let cases = FSharpType.GetUnionCases typ 
              cases.Length > 0 && equivHeadTypes (typedefof<list<_>>) cases.[0].DeclaringType)
 
-        module Type =
```
Comment (TIHan) [inline src/utils/sformat.fs]: Why was all of this deleted? Only curious! I'm just ignorant.

**Comment by TIHan** [comment]
I just did a brief scan of everything. I do have a pet peeve when I see commented code without an explanation for why something was commented out.  I put in some comments, I don't know if they are helpful. I'm no expert here, so there may be good reasons for everything and I am only trying to help when I can! Otherwise, everything seems to look fine on my end; I haven't tested anything though.

**Comment by vasily-kirichenko** [comment]
Just a side note, I've tested struct tuples in a scenario where they were supposed to speedup returning multiple values from a function http://vaskir.blogspot.ru/2016/05/upcoming-f-struct-tuples-are-they.html 
It turned out they makes no difference.

**Comment by TIHan** [comment]
@vasily-kirichenko The current example isn't a good estimate, as if you compile the first non-struct tuple example in Release, it compiles to this:
```
[CompilationMapping(SourceConstructFlags.Module)]
public static class Program
{
  [EntryPoint]
  public static int main(string[] _arg1)
  {
    Stopwatch stopwatch = Stopwatch.StartNew();
    for (int index = 1; index < 100000001; ++index)
      Math.Sin((double) index / 1000.0);
    stopwatch.Stop();
    PrintfModule.PrintFormatLineToTextWriter<FSharpFunc<TimeSpan, Unit>>(Console.Out, (PrintfFormat<FSharpFunc<TimeSpan, Unit>, TextWriter, Unit, Unit>) new PrintfFormat<FSharpFunc<TimeSpan, Unit>, TextWriter, Unit, Unit, TimeSpan>("Run: %O")).Invoke(stopwatch.Elapsed);
    stopwatch.Restart();
    GC.Collect(2);
    stopwatch.Stop();
    PrintfModule.PrintFormatLineToTextWriter<FSharpFunc<TimeSpan, Unit>>(Console.Out, (PrintfFormat<FSharpFunc<TimeSpan, Unit>, TextWriter, Unit, Unit>) new PrintfFormat<FSharpFunc<TimeSpan, Unit>, TextWriter, Unit, Unit, TimeSpan>("GC.Collect: %O")).Invoke(stopwatch.Elapsed);
    Console.ReadKey();
    return 0;
  }
}
```
It doesn't even construct the tuples in the first place.

Try this:
```
open System
open System.Diagnostics

[<EntryPoint>]
let main _ = 
    let foo x = (x, Math.Sin x)

    let mutable v = (0., 0.)

    let sw = Stopwatch.StartNew()
    for n in 1..100000000 do
        v <- foo (float n / 1000.)
        ()
    sw.Stop()
    printfn "Run: %O" sw.Elapsed
    sw.Restart()
    GC.Collect 2
    sw.Stop()
    printfn "GC.Collect: %O" sw.Elapsed
    Console.ReadKey() |> ignore
    0
```

**Comment by vasily-kirichenko** [comment]
@TIHan same time for your last snippet.

**Comment by TIHan** [comment]
Gotcha, weird I guess. I wonder if anything regressed as far as optimizing out tuples.

**Comment by TIHan** [comment]
When I use the first example @vasily-kirichenko posted, the compiler optimizes out the tuples. I didn't use this branch though. If I use the snippet I posted, I get similar timings.

**Comment by TIHan** [comment]
I also did this as a test:
```
open System
open System.Diagnostics

[<Struct>]
type StructTuple<'T1, 'T2> =

    val X : 'T1

    val Y : 'T2

    new (x, y) = { X = x; Y = y }

[<EntryPoint>]
let main _ = 
    let foo x = StructTuple (x, Math.Sin x)

    let mutable v = StructTuple (0., 0.)

    let sw = Stopwatch.StartNew()
    for n in 1..100000000 do
        v <- foo (float n / 1000.)
        ()
    sw.Stop()
    printfn "Run: %O" sw.Elapsed
    sw.Restart()
    GC.Collect 2
    sw.Stop()
    printfn "GC.Collect: %O" sw.Elapsed
    Console.ReadKey() |> ignore
    0
```
My timings were:
```
Run: 00:00:00.2693048
GC.Collect: 00:00:00.0001226
```
An order of magnitude faster.

I should just use this branch.

**Comment by vasily-kirichenko** [comment]
I added results in release mode and added decompiled code as well (in the blog post).

@dsyme `StructTuple`s are not erased in Release mode, but old `Tuple`s are. Is it intentionally?

File: src/fsharp/TastOps.fs (hunk 1 above)
Comment (dsyme) [inline src/fsharp/TastOps.fs]: I'll remove it.  It was related to [this possible extension](https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1006-struct-tuples.md#possible-extension-structness-inference)

File: src/utils/sformat.fs (hunk 5 above)
Comment (dsyme) [inline src/utils/sformat.fs]: Dead code

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

**Comment by KevinRansom** [comment]
@dsyme can you fix the merge issues please

**Comment by dsyme** [comment]
@KevinRansom I've merged.  The changes look ok.  i'll add some line notes here

File: src/fsharp/FSharp.Compiler/FSharp.Compiler.fsproj [hunk 6]
```diff
@@ -527,6 +527,7 @@
     <Reference Include="Microsoft.DiaSymReader"><HintPath>..\..\..\packages\Microsoft.DiaSymReader.1.0.8\lib\portable-net45+win8\Microsoft.DiaSymReader.dll</HintPath></Reference>
     <Reference Include="System.Reflection.Metadata"><HintPath>..\..\..\packages\System.Reflection.Metadata.1.4.1-beta-24227-04\lib\portable-net45+win8\System.Reflection.Metadata.dll</HintPath></Reference>
     <Reference Include="System.Collections.Immutable"><HintPath>..\..\..\packages\System.Collections.Immutable.1.2.0\lib\portable-net45+win8+wp8+wpa81</HintPath></Reference>
+    <Reference Include="System.ValueTuple"><HintPath>..\..\..\packages\System.ValueTuple.4.0.0-rc3-24212-01\lib\netstandard1.1</HintPath></Reference>
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler/FSharp.Compiler.fsproj]: This reference can be removed, I think it should not be needed any more now that FSharp.Core doesn't depend on System.ValueType.  At least until we deliberately use struct tuples in the compiler

File: src/fsharp/FSharp.Compiler-proto/FSharp.Compiler-proto.fsproj [hunk 7]
```diff
@@ -477,6 +477,7 @@
     <Reference Include="Microsoft.DiaSymReader"><HintPath>..\..\..\packages\Microsoft.DiaSymReader.1.0.8\lib\portable-net45+win8\Microsoft.DiaSymReader.dll</HintPath></Reference>
     <Reference Include="System.Reflection.Metadata"><HintPath>..\..\..\packages\System.Reflection.Metadata.1.4.1-beta-24227-04\lib\portable-net45+win8\System.Reflection.Metadata.dll</HintPath></Reference>
     <Reference Include="System.Collections.Immutable"><HintPath>..\..\..\packages\System.Collections.Immutable.1.2.0\lib\portable-net45+win8+wp8+wpa81</HintPath></Reference>
+    <Reference Include="System.ValueTuple"><HintPath>..\..\..\packages\System.ValueTuple.4.0.0-rc3-24212-01\lib\netstandard1.1</HintPath></Reference>
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler-proto/FSharp.Compiler-proto.fsproj]: This reference can be removed, I think it should not be needed any more now that FSharp.Core doesn't depend on System.ValueType. At least until we deliberately use struct tuples in the compiler

File: src/fsharp/FSharp.Core/FSharp.Core.fsproj [hunk 8]
```diff
@@ -213,15 +213,10 @@
...
-                                               '$(TargetFramework)' == 'XNA\5.0' OR 
-                                               '$(TargetFramework)' == 'sl3-wp' "/>
-    <Reference Include="System.Observable" Condition="'$(TargetFramework)' == 'sl3-wp' "/>
-    <Reference Include="System.Core" Condition="'$(TargetFramework)' == 'sl5' OR 
-                                                '$(TargetFramework)' == 'portable47' OR 
-                                                '$(TargetFramework)' == 'XNA\5.0' OR 
-                                                '$(TargetFramework)' == 'sl3-wp' "/>
+  </ItemGroup>
+  <ItemGroup Condition="'$(TargetFramework)' != 'portable47' and '$(TargetFramework)' != 'coreclr'">
+    <Reference Include="System.ValueTuple"><HintPath>..\..\..\packages\System.ValueTuple.4.0.0-rc3-24212-01\lib\netstandard1.1\System.ValueTuple.dll</HintPath></Reference>
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/FSharp.Core.fsproj]: This reference really needs to be removed to make sure we are not taking a static dependenct on System.ValueTuple.dll

File: src/fsharp/FSharp.Core/project.json [hunk 9]
```diff
@@ -24,7 +24,8 @@
     "System.Threading.Tasks.Parallel": "4.0.1",
     "System.Threading.Thread": "4.0.0",
     "System.Threading.ThreadPool": "4.0.10",
-    "System.Threading.Timer": "4.0.1"
+    "System.Threading.Timer": "4.0.1",
+    "System.ValueTuple": "4.0.0-rc3-24212-01",
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/project.json]: This reference can be removed,

File: src/fsharp/FSharp.LanguageService.Compiler/FSharp.LanguageService.Compiler.fsproj [hunk 10]
```diff
@@ -586,6 +586,7 @@
     <Reference Include="Microsoft.DiaSymReader"><HintPath>..\..\..\packages\Microsoft.DiaSymReader.1.0.8\lib\portable-net45+win8\Microsoft.DiaSymReader.dll</HintPath></Reference>
     <Reference Include="System.Reflection.Metadata"><HintPath>..\..\..\packages\System.Reflection.Metadata.1.4.1-beta-24227-04\lib\portable-net45+win8\System.Reflection.Metadata.dll</HintPath></Reference>
     <Reference Include="System.Collections.Immutable"><HintPath>..\..\..\packages\System.Collections.Immutable.1.2.0\lib\portable-net45+win8+wp8+wpa81</HintPath></Reference>
+    <Reference Include="System.ValueTuple"><HintPath>..\..\..\packages\System.ValueTuple.4.0.0-rc3-24212-01\lib\netstandard1.1</HintPath></Reference>
```
Comment (dsyme) [inline src/fsharp/FSharp.LanguageService.Compiler/FSharp.LanguageService.Compiler.fsproj]: This reference can be removed

**Comment by dsyme** [comment]
Awesome to see this passing green

**Comment by dsyme** [comment]
This looks very close to being ready. However I would expect we should wait until we are 100% certain of the final shape and form of the C# feature, and it has been committed and finalized.  I'd hate to have F# commit and then C# change some design detail :)

**Comment by dsyme** [comment]
This code is raising caught exceptions on every FSharp.Core load
```
#if FX_ASSEMBLYLOADBYSTRING
            let a = Assembly.Load("System.ValueTuple")
#else
            let a = Assembly.Load(new AssemblyName("System.ValueTuple"))
#endif 
```

at least if System.ValueTuple is not in the compiler directory by some chance.  But in any case it seems a bit inefficient on every library startup

The root cause is that in reflect.fs, these should be computed on-demand rather than be top-level constants, so you only pay the cost once you start to use reflection
```
    let stuple1 = reflectedValueTupleType 1 [| typedefof<obj> |]
    let stuple2 = reflectedValueTupleType 2 [| typedefof<obj>; typedefof<obj> |]
    let stuple3 = reflectedValueTupleType 3 [| typedefof<obj>; typedefof<obj>; typedefof<obj> |]
    let stuple4 = reflectedValueTupleType 4 [| typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj> |]
    let stuple5 = reflectedValueTupleType 5 [| typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj> |]
    let stuple6 = reflectedValueTupleType 6 [| typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj> |]
    let stuple7 = reflectedValueTupleType 7 [| typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj> |]
    let stuple8 = reflectedValueTupleType 8 [| typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; typedefof<obj>; deOption(reflectedValueTuple 0) |]
```

**Comment by KevinRansom** [comment]
@dsyme Thanks for this important feature.

**Comment by cartermp** [comment]
👍 👍 👍 

🎉 🎉 🎉 

😸 😸 😸

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

**Pull request merged**